### ISSUE TITLE
applying a service twice with clusterIP set to blank returns an error

### ISSUE THREAD

**Issue by deitch** [kubernetes/kubernetes#68369]
**Is this a BUG REPORT or FEATURE REQUEST?**:
BUG REPORT

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind feature

**What happened**:

1. Create a small file `service.yml` with a Service in it, and `clusterIP: ""`
2. `kubectl apply -f service.yml` - the service is created
3. `kubectl apply -f service.yml` - get an error

Then error is the classic:

```
F0906 19:32:48.426166   87883 helpers.go:119] The Service "myservice" is invalid: spec.clusterIP: Invalid value: "": field is immutable
```

I did check the annotation `kubectl.kubernetes.io/last-applied-configuration` to see that it has the `clusterIP: ""`, which it does. Nonetheless, it treats it as mutated.

In short: `kubectl apply -f service.yml` done twice should always work, since the file has not changed.

**What you expected to happen**:

The service to be unchanged.

**How to reproduce it (as minimally and precisely as possible)**:

See steps above. 

**Anything else we need to know?**:

It also occurs when the first apply does not set `clusterIP` and the second one sets it to `""` or vice-versa, see #65272 . This is more basic, though.

**Environment**:
- Kubernetes version (use `kubectl version`): tried with client 1.9.x through 1.11.x, server 1.9.x through 1.10.x
- Cloud provider or hardware configuration: tried bare metal, on EC2, and docker for mac
- OS (e.g. from /etc/os-release): linuxkit (docker for Mac), CoreOS on bare metal and EC2

**Comment by liggitt** [comment]
explicitly setting the field to "" will attempt to clear it on the second application. 

If you don't want to attempt to control that field, omit it from the manifest

**Comment by deitch** [comment]
Hey @liggitt . I put on the cli sig because when I ran it with `--v=10`, it was clear that at least some of the reconciliation was happening client-side.

**Comment by deitch** [comment]
> explicitly setting the field to "" will attempt to clear it on the second application.

Right, even if I don't change the field, just do `kubectl apply -f ` twice.

If so, then `clusterIP: ""` shouldn't be a valid setting, since it is not apply-able twice. Yet the docs say it is.

**Comment by liggitt** [comment]
> it was clear that at least some of the reconciliation was happening client-side.

Correct, the cli is submitting what is contained in your manifest. If you don't intend to control the clusterIP field, omit it from your manifest. 

> clusterIP: "" shouldn't be a valid setting, since it is not apply-able twice

That doesn't follow. Somethings are valid on creation requests and not on update requests.

**Comment by deitch** [comment]
> That doesn't follow. Some things are valid on creation requests and not on update requests.

That does make automated operation (CD pipeline) far more difficult, and makes it easy to violate the "principle of least surprise". 

What else is there like that in the API with `apply`?

**Comment by liggitt** [comment]
> That does make automated operation (CD pipeline) far more difficult, and makes it easy to violate the "principle of least surprise"

The solution is to only include the fields in your manifest you intend to set to those values.

**Comment by deitch** [comment]
> The solution is to only include the fields in your manifest you intend to set to those values.

So a git repo that has a `kubernetes/` directory with ymls to apply would have to change based on the first time it is applied by CD vs the second?

If you mean, "so never set `clusterIP: ""`, ok, but then it should not be a valid accepted field.

**Comment by liggitt** [comment]
> If you mean, "so never set clusterIP: "", ok, but then it should not be a valid accepted field.

The behavior that makes that acceptable on create and not on update is part of the service rest handler, which allocates cluster IP on creation of the service and does not permit clearing it via update. 

this behavior is not possible to change and remain backward compatible

/close

**Comment by deitch** [comment]
> The behavior that makes that acceptable on create and not on update is part of the service rest handler, which allocates cluster IP on creation of the service and does not permit clearing it via update.

Understood from the technical perspective.

From a usage perspective (the important part), it creates hurdles and makes automation difficult. At the least, I think we should deprecate the `clusterIP: ""` if you cannot use it twice.

**Comment by mans0954** [comment]
This also creates a problem using the Kubernetes API with something like Google Cloud Deployment Manager (GCDM). Minimum steps to demonstrate the problem:
# Create a minimal type provider for the GKE API

Create the file '/type-provider-minimal.jinja`:

```
resources:
- name: api-v1-tp
  type: deploymentmanager.v2beta.typeProvider
  properties:
    options:
      validationOptions:
        schemaValidation: IGNORE_WITH_WARNINGS
      inputMappings:
      - fieldName: Authorization
        location: HEADER
        value: >
          $.concat("Bearer ", $.googleOauth2AccessToken())
    descriptorUrl: https://{{ properties["k8s_endpoint"] }}/swaggerapi/api/v1
- name: apis-apps-v1-tp
  type: deploymentmanager.v2beta.typeProvider
  properties:
    options:
      validationOptions:
        schemaValidation: IGNORE_WITH_WARNINGS
      inputMappings:
      - fieldName: Authorization
        location: HEADER
        value: >
          $.concat("Bearer ", $.googleOauth2AccessToken())
    descriptorUrl: https://{{ properties["k8s_endpoint"] }}/swaggerapi/apis/apps/v1
```
and deploy it with GCDM:
```
 gcloud deployment-manager deployments create csh-type-provider-minimal --template=type-provider-minimal.jinja --properties=k8s_endpoint:XXX.XXX.XXX.XXX
```
(here XXX.XXX.XXX.XXX is the IP address of our Kubernetes master).
# Create a service

Create the file `test-svc`:
```
{% set APIS_APPS_V1_CLUSTER_TYPE = env['project'] + '/apis-apps-v1-tp' %}

{% set SERVICE_COLLECTION = '/api/v1/namespaces/{namespace}/services' %}

{% set DEPLOYMENT_COLLECTION = '/apis/apps/v1/namespaces/{namespace}/deployments' %} 

resources:
- name: test-svc
  type: {{ API_V1_CLUSTER_TYPE }}:{{ SERVICE_COLLECTION }}
  properties:
    apiVersion: v1
    kind: Service
    namespace: default
    metadata:
      name: test-svc
    spec:
      ports:
      - name: port0
        port: 8090
        protocol: TCP
        targetPort: 8080
      selector:
        app: test-app
- name: test-deployment
  type: {{ APIS_APPS_V1_CLUSTER_TYPE }}:{{ DEPLOYMENT_COLLECTION }}
  properties:
    apiVersion: apps/v1
    kind: Deployment
    namespace: default
    metadata:
      name: test-deployment
      labels:
        app: test-app
    spec:
      selector:
        matchLabels:
          app: test-app
      template:
        metadata:
          labels:
            app: test-app
        spec:
          containers:
          - image: gcr.io/hello-minikube-zero-install/hello-node
            name: test-app
```
and deploy it:
```
gcloud deployment-manager deployments create csh-test --template=test.jinja
```
Run `kubectl exec -ti <pod-name> /bin/bash` to log into a container in the default namespace and then run `curl http://test-svc:8090/` to check that `Hello World` is displayed. So far, so good.

# Updating the service

Now modify the service, e.g. copy the file `test.jinja` to `test-modified.jinja` and add an extra port to `test-svc`:
```
    spec:
      ports:
        ...
      - name: port1
        port: 8091
        protocol: TCP
        targetPort: 8080
```
Now update the service:
```
gcloud deployment-manager deployments update csh-test --template=forGoogle/test-modified.jinja
The fingerprint of the deployment is 23G7_WkD_3I4q00Ntp_aSg==
Waiting for update [operation-1549451161859-58137b619779e-cb2b19a1-71726f37]...failed.                                                                                                                                  
ERROR: (gcloud.deployment-manager.deployments.update) Error in Operation [operation-1549451161859-58137b619779e-cb2b19a1-71726f37]: errors:
- code: RESOURCE_ERROR
  location: /deployments/csh-test/resources/test-svc
  message: '{"ResourceType":"ol-stag/api-v1-tp:/api/v1/namespaces/{namespace}/services","ResourceErrorCode":"422","ResourceErrorMessage":{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Service
    \"test-svc\" is invalid: [metadata.resourceVersion: Invalid value: \"\": must
    be specified for an update, spec.clusterIP: Invalid value: \"\": field is immutable]","reason":"Invalid","details":{"name":"test-svc","kind":"Service","causes":[{"reason":"FieldValueInvalid","message":"Invalid
    value: \"\": must be specified for an update","field":"metadata.resourceVersion"},{"reason":"FieldValueInvalid","message":"Invalid
    value: \"\": must be specified for an update","field":"metadata.resourceVersion"},{"reason":"FieldValueInvalid","message":"Invalid
    value: \"\": field is immutable","field":"spec.clusterIP"}]},"code":422,"statusMessage":"{\"kind\":\"Status\",\"apiVersion\":\"v1\",\"metadata\":{},\"status\":\"Failure\",\"message\":\"Service
    \\\"test-svc\\\" is invalid: [metadata.resourceVersion: Invalid value: \\\"\\\":
    must be specified for an update, spec.clusterIP: Invalid value: \\\"\\\": field
    is immutable]\",\"reason\":\"Invalid\",\"details\":{\"name\":\"test-svc\",\"kind\":\"Service\",\"causes\":[{\"reason\":\"FieldValueInvalid\",\"message\":\"Invalid
    value: \\\"\\\": must be specified for an update\",\"field\":\"metadata.resourceVersion\"},{\"reason\":\"FieldValueInvalid\",\"message\":\"Invalid
    value: \\\"\\\": must be specified for an update\",\"field\":\"metadata.resourceVersion\"},{\"reason\":\"FieldValueInvalid\",\"message\":\"Invalid
    value: \\\"\\\": field is immutable\",\"field\":\"spec.clusterIP\"}]},\"code\":422}\n","requestPath":"https://35.231.142.169:443/api/v1/namespaces/default/services/test-svc","httpMethod":"PUT"}}'
```

This was tested against 1.11.6-gke.2 on the Kubernetes master.

@liggitt 's suggestion to omit clusterIP from the manifest doesn't directly apply here, since the null clusterIP is presumably being inserted by GCDM, presumably in response to how it interprets Kubernetes' Swagger API definition?

**Comment by klarose** [comment]
> The behavior that makes that acceptable on create and not on update is part of the service rest handler, which allocates cluster IP on creation of the service and does not permit clearing it via update.
this behavior is not possible to change and remain backward compatible

I don't quite understand how it is impossible to change this. The existing behaviour is that you cannot go from "allocated" to "unallocated" on an update. Rather than asserting on the existence of the field, why not just assert on the change in the value? How does that break compatibility? Are users relying on this case being rejected as part of their workflow? Or is it just challenging to change? I'm fine with the latter, that is different from an impossibility.

Either way, why are we asserting on the structure of the document rather than the semantics of the document? It is clear to me, and apparently quite a few others, that `ClusterIP` not being present is semantically equivalent to `ClusterIP` being equal to "".

**Comment by irvifa** [comment]
Seems like this is related to #11237

**Comment by feluxe** [comment]
I guess this is an example of how the issue unfolds in practice: https://github.com/helm/charts/issues/9227

**Comment by thockin** [comment]
"No value" is different from "the empty string" (which is a value).

TL;DR: don't set values for fields that you mean to say "no value".

**Comment by klarose** [comment]
> "No value" is different from "the empty string" (which is a value).
> 
> TL;DR: don't set values for fields that you mean to say "no value".

I think part of the problem here is that the API is changing the spec in the backend on the user, which is surprising. I think it's a fairly reasonable principle that you should have able to do apply a kubernetes resource without having to know whether it has been applied before.

I.e. if I run

```
kubectl apply -f my-service.yaml
kubectl apply -f my-service.yaml
```

the second apply should succeed as long as the first does, since I have not changed the specification of the resource.

The fact that this does not work if the clusterIP is `""` tells me either that it's a mistake for the API to accept `""` as a valid value, or that it should understand that `""` has special meaning. Certainly it does on create, so why can it not on apply?

For what it's worth, if you don't specify anything for clusterIP, it looks like the API fills it in for you, so if you leave it empty entirely, my example will work. What's the point in allowing "", then? Is it simply for backwards compatibility? If not, then does "" really mean "no value" or does it mean "choose the IP for me"? If it's "choose the IP for me", why does that not still apply on subsequent operations from the initial create?

**Comment by thockin** [comment]
> I think part of the problem here is that the API is changing the spec in
the backend on the user, which is surprising.

We have all sorts of controllers which modify spec.  For example -
autoscalers.

> The fact that this does not work if the clusterIP is "" tells me either
that it's a mistake for the API to accept "" as a valid value, or that it
should understand that "" has special meaning.

Yes.  Unfortunately, "" and "not specified" are conflated on create (an
artifact of history) and we can't really fix it.  To fix it we'd either
break all clients who pass "" in the first place or we'd hand-code PATCH
behavior, which has unknown future implications.  If I had a time machine,
I'd make "" always be an error, but to do that would be a breaking change.

> What's the point in allowing "", then?

It's not so much "allowed" as it is something that shakes out from a set of
unrelated decisions and historical context.  Service is one of the oldest
APIs in the system, so many of our current best-practices are not
applicable (again - breaking changes are not OK).

> or does it mean "choose the IP for me"? If it's "choose the IP for me",
why does that not still apply on subsequent operations from the initial
create?

If we retained that semantic after the initial allocation, it would
re-allocate a new IP every time you apply "".  That's CLEARLY not what most
users intend.  Again, "" should always have been an error.  The fact that
it is allowed at all is the implementation leaking through.

**Comment by klarose** [comment]
> If we retained that semantic after the initial allocation, it would
> re-allocate a new IP every time you apply "".  That's CLEARLY not what most
> users intend.  Again, "" should always have been an error.  The fact that
> it is allowed at all is the implementation leaking through.

Fair point. If this is still coming up, I wonder if a warning in kubectl tied to this specific resource type (if that's not too onerous design-wise) suggesting leaving it unset could help. Maybe it should also be noted in the documentation somehow.

Thanks for the detailed explanation either way.
